src/sparse/KokkosSparse_spadd.hpp

```diff
@@ -0,0 +1,733 @@
+      //run the unsorted c_rowmap upper bound functor (just adds together A and B entry counts row by row)
+      clno_row_view_t_ c_rowmap_upperbound("C row counts upper bound", nrows + 1);
+      size_t c_nnz_upperbound = 0;
+      {
+        clno_row_view_t_ c_rowcounts_upperbound("C row counts upper bound", nrows);
+        UnsortedEntriesUpperBound<size_type, alno_row_view_t_, blno_row_view_t_, clno_row_view_t_>
+          countEntries(a_rowmap, b_rowmap, c_rowcounts_upperbound);
+        Kokkos::parallel_for(range_type(0, nrows), countEntries);
+        execution_space::fence();
+        //get (temporary) c_rowmap as cumulative sum
+        parallel_prefix_sum<size_type, clno_row_view_t_> prefix(c_rowcounts_upperbound, c_rowmap_upperbound);
+        Kokkos::parallel_scan(range_type(0, nrows + 1), prefix);
+        //compute uncompressed entries of C (just indices, no scalars)
+        execution_space::fence();
+        c_nnz_upperbound = c_rowmap_upperbound(nrows);
+      }
+      clno_nnz_view_t_ c_entries_uncompressed("C entries uncompressed", c_nnz_upperbound);
+      clno_nnz_view_t_ ab_perm("A and B permuted entry indices", c_nnz_upperbound);
+      //compute the unmerged sum
+      UnmergedSumFunctor<size_type, ordinal_type, alno_row_view_t_, blno_row_view_t_, clno_row_view_t_,
+                         alno_nnz_view_t_, blno_nnz_view_t_, clno_nnz_view_t_> unmergedSum(
+                         a_rowmap, a_entries, b_rowmap, b_entries, c_rowmap_upperbound, c_entries_uncompressed, ab_perm);
+      Kokkos::parallel_for(range_type(0, nrows), unmergedSum);
+      execution_space::fence();
+      //sort the unmerged sum
+      SortEntriesFunctor<size_type, clno_row_view_t_, clno_nnz_view_t_>
+        sortEntries(c_rowmap_upperbound, c_entries_uncompressed, ab_perm);
+      Kokkos::parallel_for(range_type(0, nrows), sortEntries);
+      execution_space::fence();
+      clno_nnz_view_t_ a_pos("A entry positions", a_rowmap(nrows));
+      clno_nnz_view_t_ b_pos("B entry positions", b_rowmap(nrows));
+      //merge the entries and compute Apos/Bpos, as well as Crowcounts
+      {
+        clno_row_view_t_ c_rowcounts("C row counts", nrows);
+        MergeEntriesFunctor<size_type, ordinal_type, alno_row_view_t_, blno_row_view_t_, clno_row_view_t_, clno_nnz_view_t_>
+          mergeEntries(a_rowmap, b_rowmap, c_rowmap_upperbound, c_rowcounts, c_entries_uncompressed, ab_perm, a_pos, b_pos);
+        Kokkos::parallel_for(range_type(0, nrows), mergeEntries);
+        execution_space::fence();
+        //compute actual c_rowmap
+        parallel_prefix_sum<size_type, clno_row_view_t_> prefix(c_rowcounts, c_rowmap);
+        Kokkos::parallel_scan(range_type(0, nrows + 1), prefix);
+        execution_space::fence();
+      }
+      addHandle->set_a_b_pos(a_pos, b_pos);
+    }
+    //provide the number of NNZ in C to user through handle
+    addHandle->set_max_result_nnz(c_rowmap(nrows));
```

Comment:
This cause access issues if UVM is not enabled. Something like below is needed for this.
    auto d_c_nnz_size = Kokkos::subview(c_rowmap, nrows);
    auto h_c_nnz_size = Kokkos::create_mirror_view (d_c_nnz_size);
    Kokkos::deep_copy (h_c_nnz_size, d_c_nnz_size);
    execution_space::fence();
    addHandle->set_max_result_nnz(h_c_nnz_size());